internal/executor/commanditerator/commanditerator.go

```diff
@@ -0,0 +1,56 @@
+package commanditerator
+
+import (
+	"github.com/cirruslabs/cirrus-ci-agent/api"
+)
+
+type CommandIterator struct {
+	commands []*api.Command
+	idx int
+}
+
+func New(commands []*api.Command) *CommandIterator {
+	return &CommandIterator{
+		commands: commands,
+		idx: 0,
+	}
+}
+
+func (ci *CommandIterator) getOrPeek(failedAtLeastOnce bool, peek bool) *api.Command {
+	shadowIdx := ci.idx
+	defer func() {
+		if peek {
+			return
+		}
+
+		ci.idx = shadowIdx
+	}()
+
+	for {
+		if shadowIdx >= len(ci.commands) {
+			return nil
+		}
+
+		nextCommand := ci.commands[shadowIdx]
+
+		shadowIdx++
+
+		if shouldRun(nextCommand, failedAtLeastOnce) {
```

Comment:
This logic seems already "skips" some of the commands which we'll need to re-implement on the backend.

The idea was also to remove a retry around `client.CirrusClient.ReportSingleCommand` and have it only for the ReportAgentFinihsed. This will tolerate potential failures in case of network or some other instabilities.

I think passing a list of updates might be a slightly simpler and more concise in the implementation.

Reply:
Even with the list of updates we still need to send the next command name for the web UI to update.

Reply:
We can send update with `EXECUTING` status 🤔